### PR TITLE
work around #2703 and release pfio memory as early as possible

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
## Types of change(s)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist
- [x] Tested this change with a run of GEOSgcm
- [x] Ran the Unit Tests (`make tests`)

## Description
Save data to a integer array pointer to work around #2703 . This PR also releases the pfio memory as early as possible

### CODE REVIEW

**Comment by weiyuan-jiang** [comment]
@bena-nasa Can you test this PR can write more 2d variables ?

File: pfio/MultiGroupServer.F90 [hunk 1]
```diff
@@ -419,13 +419,15 @@ subroutine receive_output_data(this, rc)
...
 
-        if (allocated(this%buffers(back_local_rank+1)%buffer)) call MPI_Wait(this%buffers(back_local_rank+1)%request, MPI_STAT, ierror)
         call f_d_ms(collection_counter)%serialize(this%buffers(back_local_rank+1)%buffer)
+        call f_d_ms(collection_counter)%destroy(_RC)
         msg_size= size(this%buffers(back_local_rank+1)%buffer)
         call Mpi_send(msg_size,1, MPI_INTEGER, this%back_ranks(back_local_rank+1), &
              this%back_ranks(back_local_rank+1), this%server_comm, ierror)
         call Mpi_Isend(this%buffers(back_local_rank+1)%buffer, msg_size, MPI_INTEGER, this%back_ranks(back_local_rank+1), &
             this%back_ranks(back_local_rank+1), this%server_comm, this%buffers(back_local_rank+1)%request,ierror)
+        call MPI_Wait(this%buffers(back_local_rank+1)%request, MPI_STAT, ierror)
```
Comment (tclune) [inline pfio/MultiGroupServer.F90:429]: Doesn't placing the WAIT here defeat the asynchronous aspects of the communication pattern?

Comment (weiyuan-jiang) [reply pfio/MultiGroupServer.F90:429]: Yes, I believe it slows down the server. That is a tradeoff.  Maybe we should make it smarter. If the memory is not an issue, then we don't make changes here

Comment (tclune) [reply pfio/MultiGroupServer.F90:429]: Unless you can measure it for some realistic use cases, let's not slow the code down.   Asynchronicity is more important than memory footprint.

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,7 +8,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
 ### Added
-
+<<<<<<< HEAD
+- Release the pfio memory as early as possible
```
Comment (tclune) [inline CHANGELOG.md:12]: fix issue in changelog

**Review by tclune — APPROVED** [review]

**Comment by weiyuan-jiang** [comment]
Ben was able to produce 141G file with 3d variable on 181 levels of CS5760

File: pfio/IntArray.F90 [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+#define _value type (IntArray)
+#define _value_equal_defined
+
+#define _map StringIntArrayMap
+#define _iterator StringIntArrayMapIterator
+
+#define _alt
+#include "templates/map.inc"
+   
+end module pFIO_StringIntArrayMapMod
```
Comment (tclune) [inline pfio/IntArray.F90:101]: I want this PR in, so I won't request you change this now.  But I've grown to regret "sneaking" a container in at the end of a module in this fashion.  Makes it harder to find.   Splitting into a separate file is the preferred approach going forward.

**Review by tclune — APPROVED** [review]

**Review by tclune — APPROVED** [review]

**Pull request merged**